en/dev_log/ulog_file_format.md

```diff
@@ -137,51 +142,57 @@ It is possible that there are more fields appended at the end of this message in
-
-There can be an arbitrary amount of fields (minimum 1), separated by `;`.
+- `format` is a plain-text string with the following format: `message_name:field0;field1;`
+  - There can be an arbitrary amount of fields (minimum 1), separated by `;`.
 
-A field has the format: `type field_name` or `type[array_length] field_name` for arrays (only fixed size arrays are supported).
+A `field` has the format: `type field_name`, or for an array: `type[array_length] field_name` is used (only fixed size arrays are supported).
 
-`type` is one of the basic binary types or a `message_name` of another format definition (nested usage).
-A type can be used before it's defined.
-There can be arbitrary nesting but no circular dependencies.
+A `type` is one of the [basic binary types](#data-types) or a `message_name` of another format definition (nested usage).
+- A type can be used before it's defined.
+  - e.g. The message `MessageA:MessageB[2] msg_b` can come before the `MessageB:uint_8[3] data`
+- There can be arbitrary nesting but **no circular dependencies**
+  - e.g. `MessageA:MessageB[2] msg_b` & `MessageB:MessageA[4] msg_a`
 
 Some field names are special:
-- `timestamp`: every [Subscription Message](#a--subscription-message) must include a timestamp field
+- `timestamp`: every [Subscription Message](#a-subscription-message) must include a timestamp field
   - Its type can be: `uint64_t` (currently the only one used), `uint32_t`, `uint16_t` or `uint8_t`.
-  - The unit is always microseconds, except for in `uint8_t` it's milliseconds.
+  - The unit is always microseconds, except for in `uint8_t` where the unit is in milliseconds.
   - A log writer must make sure to log messages often enough to be able to detect wrap-arounds and a log reader must handle wrap-arounds (and take into account dropouts).
   - The timestamp must always be monotonic increasing for a message series with the same `msg_id`.
-- Padding: field names that start with `_padding` should not be displayed and their data must be ignored by a reader.
+- `_padding{}`: field names that start with `_padding` (e.g. `_padding[3]`) should not be displayed and their data must be ignored by a reader.
   - These fields can be inserted by a writer to ensure correct alignment.
   - If the padding field is the last field, then this field will not be logged, to avoid writing unnecessary data.
   - This means the `message_data_s.data` will be shorter by the size of the padding.
   - However the padding is still needed when the message is used in a nested definition.
 
 #### 'I': Information Message
 
+The Information message defines a dictionary type definition `key` : `value` pair for any information, including but not limited to Hardware version, Software version, Build toolchain for the software, etc.
+
 ```c
-struct message_info_s {
-  struct message_header_s header;
+struct ulog_message_info_header_s {
+  uint16_t msg_size;
+  uint8_t msg_type = 'I';
```

Comment:
So actually, it turns out that currently in PX4 the struct definitions don't use the `message_header_s` struct in the beginning of each message struct's definition. So the documentation was wrong (out of date).

So I propose putting in the actual implementation like so (only edited 2 messages to be in this format), what do you think? @hamishwillee

Reply:
> So the documentation was wrong (out of date).

See my comment above - this isn't meant to document what PX4 exactly does.

Reply:
@bkueng As an ordinary reader I did look at the introduction and assume that this document reflects what PX4 exactly does. Mostly because I assume this is a custom **PX4** file format definition. Not something abstract or from some external source.

We need to capture this in the introduction - ideally with some positioning for what this doc is/is not useful for in a PX4 context.

If it doesn't reflect PX4, where did this come from? And what use is the doc - just to get a general feel for the fields you might see?

Reply:
Yes the file format is not tied to PX4. It could be used by any project with similar requirements.

PX4 follows the file format precisely, but it may use different struct namings, definitions, etc. Things that are not relevant for the on-disk file format.